heat/cluster/spectral.py

```diff
@@ -0,0 +1,309 @@
+        w = w / ht.sqrt(ht.expand_dims(degree, axis=0))
+        L = ht.eye(S.shape, dtype=S.dtype, split=S.split, device=S.device, comm=S.comm) - w
+
+    else:
+
+        L = ht.diag(degree) - S
+
+    return L
+
+
+class Spectral:
+    def __init__(
+        self,
+        n_clusters=None,
+        gamma=1.0,
+        metric="rbf",
+        laplacian="fully_connected",
+        threshold=1.0,
+        boundary="upper",
+        normalize=True,
+        n_lanczos=300,
+        assign_labels="kmeans",
+    ):
+        """
+        Spectral clustering
+
+        Parameters
+        ----------
+        n_clusters : int, optional
+        gamma : float, default=1.0
+            Kernel coefficient sigma for 'rbf', ignored for affinity='euclidean'
+        metric : string
+            How to construct the similarity matrix.
+            ‘rbf’ : construct the similarity matrix using a radial basis function (RBF) kernel.
+            'euclidean' : construct the similarity matrix as only euclidean distance
+            ‘precomputed’ : interpret X as a precomputed similarity matrix.
+        laplacian : string
+            How to calculate the graph laplacian (affinity)
+            'fully_connected'
+            'eNeighborhood'
+        theshold : float
+            Threshold for affinity matrix if laplacian='eNeighborhood'
+            Ignorded for laplacian='fully_connected'
+        boundary : string
+            How to interpret threshold
+            'upper'
+            'lower'
+        normalize : bool, default = True
+            normalized vs. unnormalized Laplacian
+        n_lanczos : int
+            number of Lanczos iterations for Eigenvalue decomposition
+        assign_labels: str, default = 'kmeans'
+             The strategy to use to assign labels in the embedding space.
+             'kmeans'
+
+        """
+        self.n_clusters = n_clusters
+        self.n_lanczos = n_lanczos
+        self.metric = metric
+        self.gamma = gamma
+        self.normalize = normalize
+        self.laplacian = laplacian
+        self.epsilon = (threshold, boundary)
+        if assign_labels == "kmeans":
+            self._cluster = ht.cluster.KMeans(init="random", max_iter=30, tol=-1.0)
+        else:
+            raise NotImplementedError(
+                "Other Label Assignment Algorithms are currently not available"
+            )
+
+        # in-place properties
+        self._components = None
+        self._cluster_centers = None
+        self._labels = None
+
+    @property
+    def components_(self):
+        """
+        Returns
+        -------
+        ht.DNDarray, shape = (n_samples, n_lanzcos):
+            Eigenvectors of the graph Laplacian
+        """
+        return self._components
+
+    @property
+    def labels_(self):
+        """
+        Returns
+        -------
+        ht.DNDarray, shape=(n_points):
+            Labels of each point.
+        """
+        return self._labels
+
+    def fit(self, X):
```

Comment:
Ca we potentially break this function down into 5(?) sub-functions, where each of them performs one of the step marked by the comment strings (#1 Calculucation of Adjacency, #2 ...)?

Reply:
I had this before, but this requires additional memory allcoation in several places. This turned out to be inefficient, especially on one GPU overflowing the memory for large number of samples. Hence I redesigned. Will also change once laplacian directly takes the similarity matrix.